### PR TITLE
Scrolling improvements

### LINKED ISSUE #822: [ScrollBar+Slider] Clicking outside the thumb doesn't scroll
The user expects scrollbars and sliders to scroll when they're clicked outside the thumb.

image

(The green zone in the picture)

### PR DESCRIPTION
**Pull request by aelij**
* Add RepeatButton
* ScrollBar has {line,page}/{up/down} buttons
* Slider track buttons
* Fixed a bug with mouse wheel coordinates

Fixes #822

### CODE REVIEW

File: src/Avalonia.Controls/Button.cs [hunk 1]
```diff
@@ -25,7 +22,7 @@ public enum ClickMode
         /// <summary>
         /// The <see cref="Button.Click"/> event is raised when the pointer is pressed.
         /// </summary>
-        Press,
+        Press
```
Comment (grokys) [inline src/Avalonia.Controls/Button.cs]: Nit: unneeded change. These things aren't super-important but they do mess with `git blame`. If you can be bothered you can go back and undo changes like this [using git rebase](https://stackoverflow.com/a/1186549/6448).

File: src/Avalonia.Controls/Button.cs [hunk 2]
```diff
@@ -38,7 +35,7 @@ public class Button : ContentControl
         /// </summary>
         public static readonly StyledProperty<ClickMode> ClickModeProperty =
             AvaloniaProperty.Register<Button, ClickMode>(nameof(ClickMode));
-
+
```
Comment (grokys) [inline src/Avalonia.Controls/Button.cs]: Nit: added spurious whitespace.

File: src/Avalonia.Controls/Button.cs [hunk 3]
```diff
@@ -247,7 +240,8 @@ protected override void OnPointerReleased(PointerReleasedEventArgs e)
                 PseudoClasses.Remove(":pressed");
                 e.Handled = true;
 
-                if (ClickMode == ClickMode.Release && new Rect(Bounds.Size).Contains(e.GetPosition(this)))
+                if (ClickMode == ClickMode.Release && 
+                    new Rect(Bounds.Size).Contains(e.GetPosition(this)))
```
Comment (grokys) [inline src/Avalonia.Controls/Button.cs]: Nit: unneeded change

File: src/Avalonia.Controls/Button.cs [hunk 4]
```diff
@@ -349,7 +340,7 @@ private void StopListeningForDefault(IInputElement root)
         /// <summary>
         /// Raises the <see cref="Click"/> event.
         /// </summary>
-        private void RaiseClickEvent()
+        protected void RaiseClickEvent()
```
Comment (grokys) [inline src/Avalonia.Controls/Button.cs]: We already have an `OnClick` method which the comments say "Raises the Click event` except it doesn't. We should probably fix that and merge this method into that method instead.

**Review by grokys — CHANGES_REQUESTED** [review]
Looks great in general; a few small things could be tidied up, but also I've noticed something that doesn't work quite right: the large change on the scrollbar should be set to the scrollbar viewport size such that clicking the track outside the thumb should scroll by a page. At the moment it scrolls by a fixed amount.

**Comment by aelij** [comment]
> At the moment it scrolls by a fixed amount.

Actually, it already scrolls by percentage to the viewport size (see [here](https://github.com/AvaloniaUI/Avalonia/pull/1132/files#diff-7a282687ba30b84d4266419956311a8bR196). I just made the default small change 5% and the large 20% :) You're right it should be a whole page.

**Comment by grokys** [comment]
Also: there are some failing tests.

**Comment by aelij** [comment]
@grokys I force pushed an update to my branch to clean up unnecessary changes (and rebased onto master).

**Review by grokys — CHANGES_REQUESTED** [review]
Changes look good, but `Slider` seems to be broken now, at least in ControlCatalog it is for me...

**Review by grokys — APPROVED** [review]

**Comment by Sorien** [comment]
btw it's intended that scrollbar keys are focusable?

**Comment by aelij** [comment]
@Sorien no, we should fix that

**Comment by kekekeks** [comment]
Was #822 fixed by this?

**Comment by aelij** [comment]
Yep. Closed

**Pull request merged**